Replace uses of bare `assert` with `OPENSIM_ASSERT`

adamkewley opened the pull request:
Fixes issue #3531

### Brief summary of changes

- Added `OpenSim/Common/Assertion.h`
  - Defined `OPENSIM_ASSERT` and `OPENSIM_ASSERT_FRMOBJ` macros
  - One difference between these macros and `OPENSIM_THROW_IF` is that the they cannot customize which exception is thrown. This is so that there is room to (e.g.) add a `OpenSim::AssertionError` exception in the future.
  - Another difference is that the thing being tested must be positive in order to *not* throw
  - And the macros are toggled via `NDEBUG`, but there is deliberate room to add `OPENSIM_FORCE_ASSERTIONS_ON` as a cmake variable, which would allow for release builds with only OpenSim-level assertions turned on
 - Rolled out the assertions code-base wide:
   - Replaced all uses of bare `assert` with `OPENSIM_ASSERT` or `OPENSIM_ASSERT_FRMOBJ` if the assertion is made within an OpenSim object. The logic being that it's better for the error message to contain as much information as possible (`_FRMOBJ`)
    - Removed now-unused assert headers (`<cassert>` and `assert.h`)
    - Removed any commented-out assertions that were probably hanging around

### Testing I've completed

- Added `testAssertion.cpp`, which exercises whether the assertions trigger and, when they do, what they print

### Looking for feedback on...

### CHANGELOG.md (choose one)

- Updated with short explanation of change+motivation

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/opensim-org/opensim-core/3548)
<!-- Reviewable:end -->

adamkewley commented on OpenSim/Common/Assertion.h:
```diff
@@ -0,0 +1,55 @@
+#define OPENSIM_ASSERTION_H_
+
+/* ------------------------------------------------------------------------- *
+*                           OpenSim:  Exception.h                            *
```
Todo

adamkewley commented on OpenSim/Common/DebugUtilities.cpp:
```diff
@@ -38,8 +38,7 @@ void Fatal_Error(const char* msg, const char* function, const char* file,
     throw std::runtime_error(str);
-    assert(false);
-    exit(1);
+    std::terminate();
```
This is a drive-by edit, but is included here because:

- The original code makes no logical sense anyway (the thread will stop on the `throw`)
- The `assert(false)` is surplus, turns up in regex searches for bare `assert`s, and is compiler-toggleable
- The `exit` (and my `std::terminate`) are also surplus, but I wasn't sure whether something funny could be going on (e.g. disabling exceptions or something unusual), so I just replaced the `exit`, with the more C++stdlib-friendly `std::terminate()`

adamkewley commented on OpenSim/Common/Exception.cpp:
```diff
@@ -58,12 +57,6 @@ exception()
     setMessage(aMsg);
     _file = aFile;
     _line = aLine;
-
```
This is also a drive-by edit that was made because I was using regex to flag uses of `assert`, and this one is likely to be a cute developer trick that was probably used a while ago (i.e. before IDEs like Visual Studio had debugger toggles to break on throwing an exception) but has been buried here.

adamkewley commented on OpenSim/Simulation/Model/GeometryPath.cpp:
```diff
@@ -28,9 +28,11 @@
-#include <OpenSim/Simulation/Wrap/PathWrap.h>
 #include "Model.h"
 
+#include <OpenSim/Common/Assertion.h>
```
Headers rearranged so that those with a matching style are clustered+alphabetical

(normalizing the header usage in OpenSim is a separate, larger, issue)

adamkewley commented on OpenSim/Utilities/simmToOpenSim/detools.c:
```diff
@@ -809,7 +809,6 @@ XForm* get_deform_xform (DeformObject* dfm, int deformMode)
       case DE_DEFORM_END_MODE:
          return &dfm->deform_end;
    }
-/*   assert(0);*/
```
Changes in this file are drive-bys on comments containing the regex `\bassert\(` (i.e. find all uses of bare `assert`)

adamkewley commented on OpenSim/Common/Test/testAssertion.cpp:
```diff
@@ -0,0 +1,76 @@
+* limitations under the License.                                             *
+* -------------------------------------------------------------------------- */
+
+#include <OpenSim/Common/Assertion.h>
```
This test implementation isn't following OpenSim's usual patterns and, instead, uses `Catch`

We've discussed this previously (e.g. with @nickbianco ), but if it's problematic to start writing tests in `Catch` now then I can refactor

nickbianco reviewed (changes requested):
Overall looks great. Just had some minor comments, including when to use `OPENSIM_ASSERT` versus `OPENSIM_ASSERT_FRMOBJ`.

> Reviewed 44 of 49 files at r1, 4 of 4 files at r4, 2 of 2 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/opensim-org/opensim-core/3548#-:-NeK52ci5ApEnfbE2p7t:b3d2u21)* status: all files reviewed, 13 unresolved discussions (waiting on @adamkewley)

___
*[`OpenSim/Common/Assertion.h` line 42 at r5](https://reviewable.io/reviews/opensim-org/opensim-core/3548#-NeJzdBx6mbrFhQijd7t:-NeJzdBy0gQq5aXPXoKB:b-ndou0p) ([raw file](https://github.com/opensim-org/opensim-core/blob/d88052e147e2e9b52bdc0b1b4b3f095e7025a85e/OpenSim/Common/Assertion.h#L42)):*
> ```objc
> }
> 
> #define OPENSIM_ASSERT_ALWAYS(CONDITION) \
> ```

You might be planning to do this already, but these new macros would benefit from some doc comments (a la, `OpenSim::Exception`).
___
*[`OpenSim/Common/XMLDocument.cpp` line 249 at r5](https://reviewable.io/reviews/opensim-org/opensim-core/3548#-NeK0-DX07KhioQD2JAD:-NeK0-DX07KhioQD2JAE:bv56gvh) ([raw file](https://github.com/opensim-org/opensim-core/blob/d88052e147e2e9b52bdc0b1b4b3f095e7025a85e/OpenSim/Common/XMLDocument.cpp#L249)):*
> ```cpp
> 
>     // Validate >=  10500 and < latest as sanity check
>     OPENSIM_ASSERT(_documentVersion >= 10500 && _documentVersion <= LatestVersion);
> ```

`OPENSIM_ASSERT_FRMOBJ`? I'm not really sure what the best practices are for using plain `THROW`/`ASSERT` versus their `FRMOBJ` counterparts. I'll mark a few other places that seem inconsistent to me, but feel free to ignore.
___
*[`OpenSim/Common/Test/testAssertion.cpp` line 60 at r5](https://reviewable.io/reviews/opensim-org/opensim-core/3548#-NeK3nuO7Jx1WDHAOrdD:-NeK3nuO7Jx1WDHAOrdE:bvljmso) ([raw file](https://github.com/opensim-org/opensim-core/blob/d88052e147e2e9b52bdc0b1b4b3f095e7025a85e/OpenSim/Common/Test/testAssertion.cpp#L60)):*
> ```cpp
>         REQUIRE_THAT(ex.what(), Catch::Matchers::Contains("SomeAssertingFunction"));
>         REQUIRE_THAT(ex.what(), Catch::Matchers::Contains("testAssertion.cpp"));
>         REQUIRE_THAT(ex.what(), Catch::Matchers::Contains("34"));  // sorry ;)
> ```

Is this the line the assertion fails at?
___
*[`OpenSim/Moco/MocoBounds.cpp` line 48 at r5](https://reviewable.io/reviews/opensim-org/opensim-core/3548#-NeK3tcs94LWPxy6EyhZ:-NeK3tcs94LWPxy6Eyh_:b8tunsu) ([raw file](https://github.com/opensim-org/opensim-core/blob/d88052e147e2e9b52bdc0b1b4b3f095e7025a85e/OpenSim/Moco/MocoBounds.cpp#L48)):*
> ```cpp
> 
> MocoBounds::MocoBounds(const Property<double>& p) : MocoBounds() {
>     OPENSIM_ASSERT(p.size() <= 2);
> ```

`OPENSIM_ASSERT_FRMOBJ`
___
*[`OpenSim/Moco/MocoTrajectory.cpp` line 1350 at r5](https://reviewable.io/reviews/opensim-org/opensim-core/3548#-NeK3yMYCoo5Gko6gwe8:-NeK3yMYCoo5Gko6gwe9:b-af1q0v) ([raw file](https://github.com/opensim-org/opensim-core/blob/d88052e147e2e9b52bdc0b1b4b3f095e7025a85e/OpenSim/Moco/MocoTrajectory.cpp#L1350)):*
> ```cpp
>         // Trapezoidal rule for uniform grid:
>         // dt / 2 (f_0 + 2f_1 + 2f_2 + 2f_3 + ... + 2f_{N-1} + f_N)
>         OPENSIM_ASSERT(numTimes > 2);
> ```

`OPENSIM_ASSERT_FRMOBJ`?
___
*[`OpenSim/Moco/MocoCasADiSolver/MocoCasADiSolver.cpp` line 133 at r5](https://reviewable.io/reviews/opensim-org/opensim-core/3548#-NeK40lX288MyX_3-VOP:-NeK40lX288MyX_3-VOQ:b-af1q0v) ([raw file](https://github.com/opensim-org/opensim-core/blob/d88052e147e2e9b52bdc0b1b4b3f095e7025a85e/OpenSim/Moco/MocoCasADiSolver/MocoCasADiSolver.cpp#L133)):*
> ```cpp
>             // The API should make it impossible for both guessFromFile and
>             // guessFromAPI to be non-empty.
>             OPENSIM_ASSERT(m_guessFromAPI.empty());
> ```

`OPENSIM_ASSERT_FRMOBJ`?
___
*[`OpenSim/Moco/MocoGoal/MocoAccelerationTrackingGoal.cpp` line 39 at r5](https://reviewable.io/reviews/opensim-org/opensim-core/3548#-NeK42JLB8fk4J73ikp9:-NeK42JLB8fk4J73ikpA:brv6nm2) ([raw file](https://github.com/opensim-org/opensim-core/blob/d88052e147e2e9b52bdc0b1b4b3f095e7025a85e/OpenSim/Moco/MocoGoal/MocoAccelerationTrackingGoal.cpp#L39)):*
> ```cpp
>         // Should not be able to supply any two simultaneously.
>         if (get_acceleration_reference_file() != "") { // acceleration ref file
>             OPENSIM_ASSERT(m_acceleration_table.getNumColumns() == 0);
> ```

`OPENSIM_ASSERT_FRMOBJ`? (and below)
___
*[`OpenSim/Moco/tropter/TropterProblem.cpp` line 41 at r5](https://reviewable.io/reviews/opensim-org/opensim-core/3548#-NeK46_d5QwO1o8T2LlJ:-NeK46_d5QwO1o8T2LlK:b-af1q0v) ([raw file](https://github.com/opensim-org/opensim-core/blob/d88052e147e2e9b52bdc0b1b4b3f095e7025a85e/OpenSim/Moco/tropter/TropterProblem.cpp#L41)):*
> ```cpp
>     const int numDerivatives =
>             (int)tropSol.adjunct_names.size() - numMultipliers;
>     OPENSIM_ASSERT(numDerivatives >= 0);
> ```

`OPENSIM_ASSERT_FRMOBJ`?
___
*[`OpenSim/Simulation/Model/PathPoint.h` line 57 at r5](https://reviewable.io/reviews/opensim-org/opensim-core/3548#-NeK4VkG70TbM0CBJMBU:-NeK4VkG70TbM0CBJMBV:bmlpb6t) ([raw file](https://github.com/opensim-org/opensim-core/blob/d88052e147e2e9b52bdc0b1b4b3f095e7025a85e/OpenSim/Simulation/Model/PathPoint.h#L57)):*
> ```objc
>     DEPRECATED_14("Use setLocation() instead.")
>     void setLocationCoord(const SimTK::State&s, int aXYZ, double aValue) {
>         OPENSIM_ASSERT_FRMOBJ(aXYZ>=0 && aXYZ<=2);
> ```

`OPENSIM_ASSERT`? (i.e., should `FRMOBJ` be used in headers? I guess it's fine as long as the object pointer is available)
___
*[`OpenSim/Simulation/SimbodyEngine/ConstantCurvatureJoint.cpp` line 59 at r5](https://reviewable.io/reviews/opensim-org/opensim-core/3548#-NeK4YMQ7ceHhcnRfS1y:-NeK4YMR1O6bhIUHkcdS:brv6nm2) ([raw file](https://github.com/opensim-org/opensim-core/blob/d88052e147e2e9b52bdc0b1b4b3f095e7025a85e/OpenSim/Simulation/SimbodyEngine/ConstantCurvatureJoint.cpp#L59)):*
> ```cpp
>                 "or simulation to avoid this state.",
>                 pos(0), bound));
>         OPENSIM_ASSERT(false);
> ```

`OPENSIM_ASSERT_FRMOBJ`? (and below)
___
*[`OpenSim/Simulation/SimbodyEngine/Coordinate.cpp` line 55 at r5](https://reviewable.io/reviews/opensim-org/opensim-core/3548#-NeK4coI1hQu6kNAl0OK:-NeK4coI1hQu6kNAl0OL:b-af1q0v) ([raw file](https://github.com/opensim-org/opensim-core/blob/d88052e147e2e9b52bdc0b1b4b3f095e7025a85e/OpenSim/Simulation/SimbodyEngine/Coordinate.cpp#L55)):*
> ```cpp
> 
>     SimTK::Real calcValue(const SimTK::Vector& x) const override {
>         OPENSIM_ASSERT(x.size() == argumentSize);
> ```

`OPENSIM_ASSERT_FRMOBJ`?
___
*[`OpenSim/Simulation/SimbodyEngine/TransformAxis.h` line 121 at r5](https://reviewable.io/reviews/opensim-org/opensim-core/3548#-NeK4ry61rVTkSqWuIWi:-NeK4ry7EOiYzFGjxTGG:b-bcrmha) ([raw file](https://github.com/opensim-org/opensim-core/blob/d88052e147e2e9b52bdc0b1b4b3f095e7025a85e/OpenSim/Simulation/SimbodyEngine/TransformAxis.h#L121)):*
> ```objc
>     /** Get one component (0,1, or 2) of the axis vector. **/
>     double getAxis(int which) const 
>     {   OPENSIM_ASSERT_FRMOBJ(0<=which && which<=2); return getAxis()[which]; }
> ```

`OPENSIM_ASSERT`?


<!-- Sent from Reviewable.io -->

nickbianco reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/opensim-org/opensim-core/3548)* status: all files reviewed, 14 unresolved discussions (waiting on @adamkewley)

___
*[`OpenSim/Common/Assertion.cpp` line 41 at r5](https://reviewable.io/reviews/opensim-org/opensim-core/3548#-NeKR_cy1zQWmGypSUCD:-NeKR_cy1zQWmGypSUCE:b-yqnh51) ([raw file](https://github.com/opensim-org/opensim-core/blob/d88052e147e2e9b52bdc0b1b4b3f095e7025a85e/OpenSim/Common/Assertion.cpp#L41)):*
> ```cpp
>         throw Exception{failingFile, failingLine, failingFunction, failingCode};
>     }
> }
> ```

Missing new line.


<!-- Sent from Reviewable.io -->

adamkewley commented:
> You might be planning to do this already, but these new macros would benefit from some doc comments (a la, OpenSim::Exception).

:+1:

> OPENSIM_ASSERT_FRMOBJ?

I intend to use it whenever possible (incl. in headers - the only thing it changes is that it forwards `this` to the failure function if you use `_FRMOBJ`).

> OPENSIM_ASSERT? (in `TransformAxis`)

Same logic (i.e. if `_FRMOBJ` works, use it, because it can potentially put more information into the error message that a user sees)

> Is this the line the assertion fails at?

Yep - I'll clarify the comment. It's `sorry` because hard-coding a source line into a test is obviously a little bit naughty

adamkewley commented:
- [OpenSim/Common/XMLDocument.cpp line 249 at r5](https://reviewable.io/reviews/opensim-org/opensim-core/3548#-NeK0-DX07KhioQD2JAD:-NeK0-DX07KhioQD2JAE:bv56gvh) ([raw file](https://github.com/opensim-org/opensim-core/blob/d88052e147e2e9b52bdc0b1b4b3f095e7025a85e/OpenSim/Common/XMLDocument.cpp#L249)) - not an `OpenSim::Object`
- [OpenSim/Common/Test/testAssertion.cpp line 60 at r5](https://reviewable.io/reviews/opensim-org/opensim-core/3548#-NeK3nuO7Jx1WDHAOrdD:-NeK3nuO7Jx1WDHAOrdE:bvljmso) ([raw file](https://github.com/opensim-org/opensim-core/blob/d88052e147e2e9b52bdc0b1b4b3f095e7025a85e/OpenSim/Common/Test/testAssertion.cpp#L60)): fixed
- [OpenSim/Moco/MocoBounds.cpp line 48 at r5](https://reviewable.io/reviews/opensim-org/opensim-core/3548#-NeK3tcs94LWPxy6EyhZ:-NeK3tcs94LWPxy6Eyh_:b8tunsu) ([raw file](https://github.com/opensim-org/opensim-core/blob/d88052e147e2e9b52bdc0b1b4b3f095e7025a85e/OpenSim/Moco/MocoBounds.cpp#L48)): - changed
- [OpenSim/Moco/MocoTrajectory.cpp line 1350 at r5](https://reviewable.io/reviews/opensim-org/opensim-core/3548#-NeK3yMYCoo5Gko6gwe8:-NeK3yMYCoo5Gko6gwe9:b-af1q0v) ([raw file](https://github.com/opensim-org/opensim-core/blob/d88052e147e2e9b52bdc0b1b4b3f095e7025a85e/OpenSim/Moco/MocoTrajectory.cpp#L1350)): - not an `Object`
- [OpenSim/Moco/MocoCasADiSolver/MocoCasADiSolver.cpp line 133 at r5](https://reviewable.io/reviews/opensim-org/opensim-core/3548#-NeK40lX288MyX_3-VOP:-NeK40lX288MyX_3-VOQ:b-af1q0v) ([raw file](https://github.com/opensim-org/opensim-core/blob/d88052e147e2e9b52bdc0b1b4b3f095e7025a85e/OpenSim/Moco/MocoCasADiSolver/MocoCasADiSolver.cpp#L133)) - changed
- [OpenSim/Moco/MocoGoal/MocoAccelerationTrackingGoal.cpp line 39 at r5](https://reviewable.io/reviews/opensim-org/opensim-core/3548#-NeK42JLB8fk4J73ikp9:-NeK42JLB8fk4J73ikpA:brv6nm2) ([raw file](https://github.com/opensim-org/opensim-core/blob/d88052e147e2e9b52bdc0b1b4b3f095e7025a85e/OpenSim/Moco/MocoGoal/MocoAccelerationTrackingGoal.cpp#L39)): - changed
- [OpenSim/Moco/tropter/TropterProblem.cpp line 41 at r5](https://reviewable.io/reviews/opensim-org/opensim-core/3548#-NeK46_d5QwO1o8T2LlJ:-NeK46_d5QwO1o8T2LlK:b-af1q0v) ([raw file](https://github.com/opensim-org/opensim-core/blob/d88052e147e2e9b52bdc0b1b4b3f095e7025a85e/OpenSim/Moco/tropter/TropterProblem.cpp#L41)): not an `Object` (`tropter::Problem`)
- [OpenSim/Simulation/Model/PathPoint.h line 57 at r5](https://reviewable.io/reviews/opensim-org/opensim-core/3548#-NeK4VkG70TbM0CBJMBU:-NeK4VkG70TbM0CBJMBV:bmlpb6t) ([raw file](https://github.com/opensim-org/opensim-core/blob/d88052e147e2e9b52bdc0b1b4b3f095e7025a85e/OpenSim/Simulation/Model/PathPoint.h#L57)): - should remain as `OPENSIM_ASSERT_FRMOBJ`, there's no downside apart from an extra `this` ptr to a fully-constructed base class
- [OpenSim/Simulation/SimbodyEngine/ConstantCurvatureJoint.cpp line 59 at r5](https://reviewable.io/reviews/opensim-org/opensim-core/3548#-NeK4YMQ7ceHhcnRfS1y:-NeK4YMR1O6bhIUHkcdS:brv6nm2) ([raw file](https://github.com/opensim-org/opensim-core/blob/d88052e147e2e9b52bdc0b1b4b3f095e7025a85e/OpenSim/Simulation/SimbodyEngine/ConstantCurvatureJoint.cpp#L59)): - changed
- [OpenSim/Simulation/SimbodyEngine/Coordinate.cpp line 55 at r5](https://reviewable.io/reviews/opensim-org/opensim-core/3548#-NeK4coI1hQu6kNAl0OK:-NeK4coI1hQu6kNAl0OL:b-af1q0v) ([raw file](https://github.com/opensim-org/opensim-core/blob/d88052e147e2e9b52bdc0b1b4b3f095e7025a85e/OpenSim/Simulation/SimbodyEngine/Coordinate.cpp#L55)): not an `Object` (is a `SimTK::Function_<T>`)
- [OpenSim/Simulation/SimbodyEngine/TransformAxis.h line 121 at r5](https://reviewable.io/reviews/opensim-org/opensim-core/3548#-NeK4ry61rVTkSqWuIWi:-NeK4ry7EOiYzFGjxTGG:b-bcrmha) ([raw file](https://github.com/opensim-org/opensim-core/blob/d88052e147e2e9b52bdc0b1b4b3f095e7025a85e/OpenSim/Simulation/SimbodyEngine/TransformAxis.h#L121)): leave as `OPENSIM_ASSERT_FRMOBJ` (it's an `Object`)

adamkewley reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/opensim-org/opensim-core/3548)* status: 40 of 50 files reviewed, 14 unresolved discussions (waiting on @nickbianco)

___
*[`OpenSim/Common/Assertion.h` line 42 at r5](https://reviewable.io/reviews/opensim-org/opensim-core/3548#-NeJzdBx6mbrFhQijd7t:-NeMt5whDbHE_jTVvHqZ:b-896fix) ([raw file](https://github.com/opensim-org/opensim-core/blob/d88052e147e2e9b52bdc0b1b4b3f095e7025a85e/OpenSim/Common/Assertion.h#L42)):*
<details><summary><i>Previously, nickbianco (Nicholas Bianco) wrote…</i></summary><blockquote>

You might be planning to do this already, but these new macros would benefit from some doc comments (a la, `OpenSim::Exception`).
</blockquote></details>

Done.
___
*[`OpenSim/Common/Assertion.cpp` line 41 at r5](https://reviewable.io/reviews/opensim-org/opensim-core/3548#-NeKR_cy1zQWmGypSUCD:-NeMtVpa152f697AO4-n:b-896fix) ([raw file](https://github.com/opensim-org/opensim-core/blob/d88052e147e2e9b52bdc0b1b4b3f095e7025a85e/OpenSim/Common/Assertion.cpp#L41)):*
<details><summary><i>Previously, nickbianco (Nicholas Bianco) wrote…</i></summary><blockquote>

Missing new line.
</blockquote></details>

Done.
___
*[`OpenSim/Common/XMLDocument.cpp` line 249 at r5](https://reviewable.io/reviews/opensim-org/opensim-core/3548#-NeK0-DX07KhioQD2JAD:-NeMtkwR7jiHawp4_aPf:b-w5l0ss) ([raw file](https://github.com/opensim-org/opensim-core/blob/d88052e147e2e9b52bdc0b1b4b3f095e7025a85e/OpenSim/Common/XMLDocument.cpp#L249)):*
<details><summary><i>Previously, nickbianco (Nicholas Bianco) wrote…</i></summary><blockquote>

`OPENSIM_ASSERT_FRMOBJ`? I'm not really sure what the best practices are for using plain `THROW`/`ASSERT` versus their `FRMOBJ` counterparts. I'll mark a few other places that seem inconsistent to me, but feel free to ignore.
</blockquote></details>

It is not an `OpenSim::Object`
___
*[`OpenSim/Common/Test/testAssertion.cpp` line 60 at r5](https://reviewable.io/reviews/opensim-org/opensim-core/3548#-NeK3nuO7Jx1WDHAOrdD:-NeMtop2-mwUyx0nLhz2:b-quajiv) ([raw file](https://github.com/opensim-org/opensim-core/blob/d88052e147e2e9b52bdc0b1b4b3f095e7025a85e/OpenSim/Common/Test/testAssertion.cpp#L60)):*
<details><summary><i>Previously, nickbianco (Nicholas Bianco) wrote…</i></summary><blockquote>

Is this the line the assertion fails at?
</blockquote></details>

Updated with explanation
___
*[`OpenSim/Moco/MocoBounds.cpp` line 48 at r5](https://reviewable.io/reviews/opensim-org/opensim-core/3548#-NeK3tcs94LWPxy6EyhZ:-NeMtrhf3zKA5ml_pINd:bb6un1j) ([raw file](https://github.com/opensim-org/opensim-core/blob/d88052e147e2e9b52bdc0b1b4b3f095e7025a85e/OpenSim/Moco/MocoBounds.cpp#L48)):*
<details><summary><i>Previously, nickbianco (Nicholas Bianco) wrote…</i></summary><blockquote>

`OPENSIM_ASSERT_FRMOBJ`
</blockquote></details>

Done
___
*[`OpenSim/Moco/MocoTrajectory.cpp` line 1350 at r5](https://reviewable.io/reviews/opensim-org/opensim-core/3548#-NeK3yMYCoo5Gko6gwe8:-NeMtt8i4-zPBSgQuItH:bricpel) ([raw file](https://github.com/opensim-org/opensim-core/blob/d88052e147e2e9b52bdc0b1b4b3f095e7025a85e/OpenSim/Moco/MocoTrajectory.cpp#L1350)):*
<details><summary><i>Previously, nickbianco (Nicholas Bianco) wrote…</i></summary><blockquote>

`OPENSIM_ASSERT_FRMOBJ`?
</blockquote></details>

`OpenSim::MocoTrajectory` is not an `OpenSim::Object`
___
*[`OpenSim/Moco/MocoCasADiSolver/MocoCasADiSolver.cpp` line 133 at r5](https://reviewable.io/reviews/opensim-org/opensim-core/3548#-NeK40lX288MyX_3-VOP:-NeMu1gj1XIm2t0wrkOM:bb6un1j) ([raw file](https://github.com/opensim-org/opensim-core/blob/d88052e147e2e9b52bdc0b1b4b3f095e7025a85e/OpenSim/Moco/MocoCasADiSolver/MocoCasADiSolver.cpp#L133)):*
<details><summary><i>Previously, nickbianco (Nicholas Bianco) wrote…</i></summary><blockquote>

`OPENSIM_ASSERT_FRMOBJ`?
</blockquote></details>

Done
___
*[`OpenSim/Moco/MocoGoal/MocoAccelerationTrackingGoal.cpp` line 39 at r5](https://reviewable.io/reviews/opensim-org/opensim-core/3548#-NeK42JLB8fk4J73ikp9:-NeMu2glBpVYHxrycxXE:bb6un1j) ([raw file](https://github.com/opensim-org/opensim-core/blob/d88052e147e2e9b52bdc0b1b4b3f095e7025a85e/OpenSim/Moco/MocoGoal/MocoAccelerationTrackingGoal.cpp#L39)):*
<details><summary><i>Previously, nickbianco (Nicholas Bianco) wrote…</i></summary><blockquote>

`OPENSIM_ASSERT_FRMOBJ`? (and below)
</blockquote></details>

Done
___
*[`OpenSim/Moco/tropter/TropterProblem.cpp` line 41 at r5](https://reviewable.io/reviews/opensim-org/opensim-core/3548#-NeK46_d5QwO1o8T2LlJ:-NeMu3nWERN6Fi4Zyofh:bewltwz) ([raw file](https://github.com/opensim-org/opensim-core/blob/d88052e147e2e9b52bdc0b1b4b3f095e7025a85e/OpenSim/Moco/tropter/TropterProblem.cpp#L41)):*
<details><summary><i>Previously, nickbianco (Nicholas Bianco) wrote…</i></summary><blockquote>

`OPENSIM_ASSERT_FRMOBJ`?
</blockquote></details>

Not an `OpenSim::Object`
___
*[`OpenSim/Simulation/Model/PathPoint.h` line 57 at r5](https://reviewable.io/reviews/opensim-org/opensim-core/3548#-NeK4VkG70TbM0CBJMBU:-NeMu7fREy1OkBNPiftZ:b-hmjns6) ([raw file](https://github.com/opensim-org/opensim-core/blob/d88052e147e2e9b52bdc0b1b4b3f095e7025a85e/OpenSim/Simulation/Model/PathPoint.h#L57)):*
<details><summary><i>Previously, nickbianco (Nicholas Bianco) wrote…</i></summary><blockquote>

`OPENSIM_ASSERT`? (i.e., should `FRMOBJ` be used in headers? I guess it's fine as long as the object pointer is available)
</blockquote></details>

Done - and it doesn't matter that it's in a header, because it's just an extra pointer arg (header discipline notwithstanding, because OpenSim hasn't got any :D)
___
*[`OpenSim/Simulation/SimbodyEngine/ConstantCurvatureJoint.cpp` line 59 at r5](https://reviewable.io/reviews/opensim-org/opensim-core/3548#-NeK4YMQ7ceHhcnRfS1y:-NeMuGYpAM2z84rEUKnX:bb6un1j) ([raw file](https://github.com/opensim-org/opensim-core/blob/d88052e147e2e9b52bdc0b1b4b3f095e7025a85e/OpenSim/Simulation/SimbodyEngine/ConstantCurvatureJoint.cpp#L59)):*
<details><summary><i>Previously, nickbianco (Nicholas Bianco) wrote…</i></summary><blockquote>

`OPENSIM_ASSERT_FRMOBJ`? (and below)
</blockquote></details>

Done
___
*[`OpenSim/Simulation/SimbodyEngine/Coordinate.cpp` line 55 at r5](https://reviewable.io/reviews/opensim-org/opensim-core/3548#-NeK4coI1hQu6kNAl0OK:-NeMuHcQ3Sbj-6RDTYIy:bewltwz) ([raw file](https://github.com/opensim-org/opensim-core/blob/d88052e147e2e9b52bdc0b1b4b3f095e7025a85e/OpenSim/Simulation/SimbodyEngine/Coordinate.cpp#L55)):*
<details><summary><i>Previously, nickbianco (Nicholas Bianco) wrote…</i></summary><blockquote>

`OPENSIM_ASSERT_FRMOBJ`?
</blockquote></details>

Not an `OpenSim::Object`
___
*[`OpenSim/Simulation/SimbodyEngine/TransformAxis.h` line 121 at r5](https://reviewable.io/reviews/opensim-org/opensim-core/3548#-NeK4ry61rVTkSqWuIWi:-NeMuL2B5xw6hiI1Yom_:b-w319q5) ([raw file](https://github.com/opensim-org/opensim-core/blob/d88052e147e2e9b52bdc0b1b4b3f095e7025a85e/OpenSim/Simulation/SimbodyEngine/TransformAxis.h#L121)):*
<details><summary><i>Previously, nickbianco (Nicholas Bianco) wrote…</i></summary><blockquote>

`OPENSIM_ASSERT`?
</blockquote></details>

Is an `Object`, so should use `FRMOBJ`


<!-- Sent from Reviewable.io -->

adamkewley reviewed (commented):
> Reviewed 38 of 49 files at r1, 2 of 4 files at r4, 1 of 2 files at r5, 6 of 6 files at r6, 3 of 3 files at r7.
> *[Reviewable](https://reviewable.io/reviews/opensim-org/opensim-core/3548)* status: 49 of 50 files reviewed, 13 unresolved discussions (waiting on @nickbianco)



<!-- Sent from Reviewable.io -->

nickbianco reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 1 files at r8, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/opensim-org/opensim-core/3548#-:-NeOahy84RBIh9o297VR:bnfp4nl)* status: all files reviewed, 1 unresolved discussion (waiting on @adamkewley)

___
*[`OpenSim/Simulation/Model/PathPoint.h` line 57 at r5](https://reviewable.io/reviews/opensim-org/opensim-core/3548#-NeK4VkG70TbM0CBJMBU:-NeOaFqx8BBLzB-PKqUQ:b-tlzzpq) ([raw file](https://github.com/opensim-org/opensim-core/blob/d88052e147e2e9b52bdc0b1b4b3f095e7025a85e/OpenSim/Simulation/Model/PathPoint.h#L57)):*
> header discipline notwithstanding, because OpenSim hasn't got any :dagger: 
:sweat_smile:  


<!-- Sent from Reviewable.io -->

nickbianco reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/opensim-org/opensim-core/3548)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @adamkewley)



<!-- Sent from Reviewable.io -->
